ddl/partition.go

```diff
@@ -0,0 +1,129 @@
+func buildCreateTablePartitionInfo(ctx sessionctx.Context, d *ddl, s *ast.CreateTableStmt, cols []*table.Column) (*model.PartitionInfo, error) {
+	if s.Partition == nil {
+		return nil, nil
+	}
+	pi := &model.PartitionInfo{
+		Type:   s.Partition.Tp,
+		Enable: ctx.GetSessionVars().EnableTablePartition,
+	}
+	if s.Partition.Expr != nil {
+		buf := new(bytes.Buffer)
+		s.Partition.Expr.Format(buf)
+		pi.Expr = buf.String()
+		if s.Partition.Tp == model.PartitionTypeRange {
+			for _, col := range cols {
+				name := strings.Replace(col.Name.String(), ".", "`.`", -1)
+				if _, ok := s.Partition.Expr.(*ast.ColumnNameExpr); ok {
+					// TODO: check that the expression returns an integer.
+				}
+				if _, ok := s.Partition.Expr.(ast.ExprNode); ok {
+					if col.Tp != mysql.TypeLong && col.Tp != mysql.TypeLonglong && fmt.Sprintf("`%s`", name) == pi.Expr {
+						return nil, errors.Trace(ErrNotAllowedTypeInPartition.GenByArgs(pi.Expr))
+					}
+				}
+			}
+		}
+	} else if s.Partition.ColumnNames != nil {
+		pi.Columns = make([]model.CIStr, 0, len(s.Partition.ColumnNames))
+		for _, cn := range s.Partition.ColumnNames {
+			pi.Columns = append(pi.Columns, cn.Name)
+		}
+	}
+	for _, def := range s.Partition.Definitions {
+		// TODO: generate multiple global ID for paritions.
+		pid, err := d.genGlobalID()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		piDef := model.PartitionDefinition{
+			Name:    def.Name,
+			ID:      pid,
+			Comment: def.Comment,
+		}
+		buf := new(bytes.Buffer)
+		for _, expr := range def.LessThan {
+			expr.Format(buf)
+			piDef.LessThan = append(piDef.LessThan, buf.String())
+			buf.Reset()
+		}
+		pi.Definitions = append(pi.Definitions, piDef)
+	}
+	return pi, nil
+}
+
+func checkCreatePartitionNameUnique(part *model.PartitionInfo) error {
+	if part.Type != model.PartitionTypeRange {
+		return nil
+	}
+	newPars := part.Definitions
+	set := make(map[string]struct{})
+	for _, newPar := range newPars {
+		if _, ok := set[strings.ToLower(newPar.Name)]; ok {
+			return ErrSameNamePartition.GenByArgs(newPar.Name)
+		}
+		set[strings.ToLower(newPar.Name)] = struct{}{}
+	}
+	return nil
+}
+
+// checkCreatePartitionValue check whether `less than value` is strictly increasing for each partition.
+func checkCreatePartitionValue(part *model.PartitionInfo) error {
+	if part.Type == model.PartitionTypeRange {
+		defs := part.Definitions
+		rangeValue := defs[0].LessThan[0]
+		if strings.EqualFold(rangeValue, "MAXVALUE") && len(defs) > 1 {
+			return errors.Trace(ErrPartitionMaxvalue)
+		}
+		currentRangeValue, err := strconv.Atoi(rangeValue)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if strings.EqualFold(defs[len(defs)-1].LessThan[0], "MAXVALUE") {
+			defs = defs[:len(defs)-1]
+		}
+
+		for i := 1; i < len(defs); i++ {
+			if strings.EqualFold(defs[i].LessThan[0], "MAXVALUE") && i != len(defs)-1 {
+				return errors.Trace(ErrPartitionMaxvalue)
+			}
+			nextRangeValue, err := strconv.Atoi(defs[i].LessThan[0])
```

Comment:
I think `nextRangeValue` is currentRangeValue, and `currentRangeValue` is prevRangeValue.